### ISSUE TITLE
On Failure should return the Failure Sealed class  

### ISSUE THREAD

**Issue by CarlosEsco** [skydoves/sandwich#56]
Since 1.2.2 on failure returns the actual message for example

```[ApiResponse.Failure.Error-BadRequest](errorResponse=Response{protocol=h2, code=400, message=, url=urlItriedtodosomethingWith})```

It would be nice if this was the actual sealed class instead of the message.  This would allow a user to just use onFailure and check the Exception/Error types themselves or do whatever with the actual sealed object.  As it stands currently you lose information if it was an error, and logging from this string requires a bunch of parsing.

For example I have an extension function that logs based off the ApiResponse<*> type.  

```kotlin
fun ApiResponse<*>.log(type: String) {
    when (this) {
        is ApiResponse.Failure.Exception -> {
            XLog.enableStackTrace(10).e("Exception $type ${this.message}", this.exception)
        }
        is ApiResponse.Failure.Error -> {
            XLog.e("error $type ${this.errorBody?.string()}")
            XLog.e("error response code ${this.statusCode.code}")
        }
        else -> {
            XLog.e("error $type")
        }
    }
}
```
I would like to do something like 

```kotlin
  val loginResponseDto = authService.login(loginRequest).onFailure { 
                this.log("Trying to login")
            }
```

but currently have to do something like this

```kotlin
  val loginResponseDto = authService.login(loginRequest).onError{ 
                this.log("Trying to login")
            }.onException{ 
                this.log("Trying to login")
            }
```

**Comment by skydoves** [comment]
Hi, you can use the `onFailure` or `suspendOnFailure` extensions as you mention above:

```kotlin
.onFailure { message ->
    log("$message")
}.suspendOnFailure { message ->
    log("$message")
}
```

**Comment by skydoves** [comment]
Hi, so do you mean it would be good if the `onFailure` extension takes the actual instance instead of the message like the other extensions?

**Comment by CarlosEsco** [comment]
I believe so.  Using my example above. 
the OnError{}  provides the ApiResponse.Failure.Error<LoginResponseDto>
the OnException{}  provides the ApiResponse.Failure.Exception<LoginResponseDto>

it would be nice if 
OnFailure{} provided ApiResponse.Failure  so then the user could decide if they want the message string, or handle the actual error/exception response directly.

**Comment by skydoves** [comment]
Thank you for the details. It's available on the new snapshot version `1.2.5-SNAPSHOT`. If you feel it's good, I will release the next version very soon. Thanks!

**Comment by CarlosEsco** [comment]
Yep this works!
